### PR TITLE
Bug 2060329: show Limit exceeded state for large number of nodes in topology

### PR DESCRIPTION
**Pull request by sahil143**
**Fixes**: https://issues.redhat.com/browse/OCPBUGSM-41289

**Analysis / Root cause**: topology view lags with large number of nodes

**Solution Description**: if number of filtered nodes exceed 100 we will show a LimitExceeded state which suggests user to use the search page 

**Screen shots / Gifs for design review**: 

image

Kapture 2022-04-13 at 20 54 07

**Unit test coverage report**: 

**Test setup:**

**Browser conformance**: 

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/topology/src/components/page/LimitExceededState.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        {t(`topology~We're sorry, but we can't support large number of resources on topology at this time. Our bots
+        are working relentlessly to improve the performance.`)}
+      </EmptyStateBody>
+      <Button variant="primary" onClick={() => history.push(`/search-page/ns/${namespace}`)}>
+        {t('topology~View resources on Search page')}
+      </Button>
+      <EmptyStateSecondaryActions>
+        <Button variant="link" onClick={onShowTopologyAnyway}>
+          {t('topology~Show topology anyway')}
+        </Button>
```
Comment (sahil143) [inline frontend/packages/topology/src/components/page/LimitExceededState.tsx:36]: Dummy text is used here. Need UX/PM input on the final text for the state.

**Review by jerolimov — CHANGES_REQUESTED** [review]
Please define a const for the current max number of nodes. We can expect that we will change it over time.

File: frontend/packages/topology/src/components/page/LimitExceededState.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        {t(`topology~Limit Exceeded`)}
+      </Title>
+      <EmptyStateBody>
+        {t(
+          `topology~We're sorry, but we can't support large number of resources on topology at this time. Our bots are working relentlessly to improve the performance.`,
+        )}
+      </EmptyStateBody>
+      <Button variant="primary" onClick={() => history.push(`/search-page/ns/${namespace}`)}>
+        {t('topology~View resources on Search page')}
+      </Button>
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/page/LimitExceededState.tsx:33]: Use semantic HTML. Navigational links shouldn't be buttons that push to history. Use react-router with the button. See patternfly docs for example code.

File: frontend/packages/topology/src/components/page/TopologyView.tsx [hunk 3]
```diff
@@ -320,25 +322,29 @@ export const ConnectedTopologyView: React.FC<ComponentProps> = ({
     }
   }, [searchParams, labelParams]);
 
+  const nodesLength = filteredModel?.nodes?.length;
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/page/TopologyView.tsx:325]:
```suggestion
  const nodesLength = filteredModel?.nodes?.length ?? 0;
```

File: frontend/packages/topology/src/components/page/TopologyView.tsx [hunk 4]
```diff
@@ -320,25 +322,29 @@ export const ConnectedTopologyView: React.FC<ComponentProps> = ({
...
-        <Topology
-          model={filteredModel}
-          namespace={namespace}
-          application={applicationRef.current}
-          onSelect={onSelect}
-          setVisualization={setVisualization}
-        />
-      ),
-    [filteredModel, namespace, onSelect, viewType],
+      nodesLength && (nodesLength < 100 || (nodesLength > 99 && showTopologyAnyway)) ? (
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/page/TopologyView.tsx:329]:
```suggestion
      nodesLength < 100 || showTopologyAnyway ? (
```

Use a const for the max nodes.

Also, do we really want to render `null` if 0 nodes? old code didn't

File: frontend/packages/topology/src/components/page/TopologyView.tsx [hunk 5]
```diff
@@ -386,7 +392,13 @@ export const ConnectedTopologyView: React.FC<ComponentProps> = ({
                         </span>
                       </div>
                     )}
-                    {viewContent}
+                    {nodesLength && nodesLength > 99 && !showTopologyAnyway ? (
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/page/TopologyView.tsx:395]: Use a const for max nodes.
```suggestion
                    {nodesLength >= MAX_NODES && !showTopologyAnyway ? (
```

File: frontend/packages/topology/src/components/page/LimitExceededState.tsx (hunk 1 above)
Comment (sahil143) [reply frontend/packages/topology/src/components/page/LimitExceededState.tsx]: /cc @serenamarie125 @beaumorley 

Need proper text for this state.

File: frontend/packages/topology/src/components/page/TopologyView.tsx [hunk 6]
```diff
@@ -320,25 +323,29 @@ export const ConnectedTopologyView: React.FC<ComponentProps> = ({
...
-        <Topology
-          model={filteredModel}
-          namespace={namespace}
-          application={applicationRef.current}
-          onSelect={onSelect}
-          setVisualization={setVisualization}
-        />
-      ),
-    [filteredModel, namespace, onSelect, viewType],
+      nodesLength < MAX_NODES_LIMIT || showTopologyAnyway ? (
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/page/TopologyView.tsx:330]: you use `nodesLength > MAX_NODES_LIMIT` when checking if the limit is exceeded, therefore you need to use `<=` here

**Review by serenamarie125 — CHANGES_REQUESTED** [review]
@sahil143 - worked with Beau and Allison Wolfe (content writer).  Please change to the following recommendations:

**Loading is taking longer than expected**

We noticed that it is taking a long time to visualize your application Topology. You can use Search to find specific resources or click Continue to keep waiting.

[Go to Search]
[Continue]

**Review by jerolimov — CHANGES_REQUESTED** [review]
@sahil143 Can you please also check the dark mode of this?

The empty state looks like this:

image

But your empty state looks like this:

image

You might want to take a look into @vikram-raj PR #11393 that solves this for the empty state above.

**Comment by sahil143** [comment]
@jerolimov fixed dark mode and added icon for the state
image

**Review by jerolimov — APPROVED** [review]
Change works fine and contains the recommendation text from Serena.

/lgtm
/approve

**Pull request merged**